sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp

```diff
@@ -509,20 +503,28 @@ cdef class EuclideanArgKmin{{name_suffix}}(ArgKmin{{name_suffix}}):
         # which keep tracks of the argkmin.
         for i in range(n_X):
             for j in range(n_Y):
-                heap_push(
-                    heaps_r_distances + i * self.k,
-                    heaps_indices + i * self.k,
-                    self.k,
-                    # Using the squared euclidean distance as the rank-preserving distance:
-                    #
-                    #             ||X_c_i||² - 2 X_c_i.Y_c_j^T + ||Y_c_j||²
+                sqeuclidean_dist_i_j = (
+                    self.X_norm_squared[i + X_start] +
+                    dist_middle_terms[i * n_Y + j] +
+                    self.Y_norm_squared[j + Y_start]
+                )
+
+                sqeuclidean_dist_i_j = (
+                    sqeuclidean_dist_i_j
+                    # Catastrophic cancellation causing -0. and NaN can be present
```

Comment:
How does nan appear when computing `sqeuclidean_dist_i_j`?

Also the `main` does not check for `nan`, was this a bug in `main`?

Reply:
Hum, thinking again, NaNs might only appear when we `sqrt` negative distances (hence the absence of the NaN guard on main).